### PR TITLE
Support customized Cid generation

### PR DESCRIPTION
**Pull request by liwenjieQu**
Currently Quinn generates cid using random number only. Some use case may require a more sophisticated cid structure. This change mainly add a `ConnectionIdGenerator` trait that allows to use any customized cid generator.

### CODE REVIEW

File: quinn-proto/src/config.rs
Comment (Ralith) [inline quinn-proto/src/config.rs:327]: I'd rather this be a setter than an additional constructor, as setters are more scalable/composable.

**Review by Ralith — COMMENTED** [review]
Thanks for the PR! I agree that we should expose something like this, e.g. to support load balancers. Out of curiosity, are you doing something with Quinn that requires this in practice?

File: quinn-proto/src/config.rs
Comment (Matthias247) [reply quinn-proto/src/config.rs:327]: The downside is that one would need a default initialization of the factory before (e.g. with the random generator factory), which would then get replaced. Introducing an `Option` and then having to deal later on with the field not being set seems not very convenient.

A `Builder` for the config could avoid this, but that's a lot bigger of an API change. 
Since the cost for the wasted `Box<RandomIdConnectionGeneratorFactory>` is likely negligible compared to the remaing setup (e.g. loading TLS keys) it's likely find to just have the setter.

File: quinn-proto/src/shared.rs
Comment (Matthias247) [inline quinn-proto/src/shared.rs:136]: Should have a doc comment since its public

File: quinn-proto/src/shared.rs
Comment (Matthias247) [inline quinn-proto/src/shared.rs:145]: doc comment

File: quinn-proto/src/shared.rs
Comment (Matthias247) [inline quinn-proto/src/shared.rs:126]: Is the `shared` module actually `pub`lically exported? Otherwise won't be usable 😀
Btw: It could also live an a seperate `connection_id_generator` module, or just be reexported from the top level module.

File: quinn-proto/src/config.rs
Comment (Ralith) [reply quinn-proto/src/config.rs:327]: In the (indeed unlikely) event that the cost became an issue, we could even use a ZST for the default impl, which wouldn't so much as malloc.

Comment (liwenjieQu) [reply quinn-proto/src/config.rs:327]: I agree. This only occurs at the configuration phase and thus the waste is not a big problem - Will proceed the change to setter.

File: quinn-proto/src/shared.rs
Comment (liwenjieQu) [reply quinn-proto/src/shared.rs:126]: Will do. This sounds good as the customized cid generator implementation can be done in a different crate and later imported into quinn-proto.

File: quinn-proto/src/lib.rs
Comment (Matthias247) [inline quinn-proto/src/lib.rs:64]: Nit: Might swap this with line 63 and have an empty line after it for consistent style

File: quinn-proto/src/config.rs
Comment (Ralith) [inline quinn-proto/src/config.rs:325]:
```suggestion
    /// Supply a custom connection ID generator factory
    ///
    /// Called once by each `Endpoint` constructed from this configuration to obtain the CID generator which will
    /// be used to generate the CIDs used for incoming packets on all connections involving that  `Endpoint`. A
    /// custom CID generator allows applications to embed information in local connection IDs, e.g. to support
    /// stateless packet-level load balancers.
```

File: quinn-proto/src/config.rs
Comment (Ralith) [inline quinn-proto/src/config.rs:331]:
```suggestion
    pub fn cid_generator<F: Fn() -> Box<dyn ConnectionIdGenerator> + Send + Sync + 'static>(
```
For concision, we omit the `set_` prefix from methods on builders where ~all methods are setters.

File: quinn-proto/src/endpoint.rs
Comment (Ralith) [inline quinn-proto/src/endpoint.rs:810]:
```suggestion
    /// Try using longer connection IDs.
```

File: quinn-proto/src/lib.rs
Comment (Ralith) [inline quinn-proto/src/lib.rs:66]: Stray newline.
```suggestion
mod token;
```

File: quinn-proto/src/shared.rs
Comment (Ralith) [inline quinn-proto/src/shared.rs:4]: `use` statements referring to external items should be in a separate block, preceding those referring to internal ones.

File: quinn-proto/src/shared.rs
Comment (Ralith) [inline quinn-proto/src/shared.rs:65]: Would it be sufficient to instead export `ConnectionId::new` and otherwise rely on the pre-existing `Deref` and `DerefMut` impls?

File: quinn-proto/src/cid_generator.rs
Comment (Ralith) [inline quinn-proto/src/cid_generator.rs:11]: It looks like this is only used to skip checking an `Endpoint`'s hash table for a CID, but that hash table should return `None` for invalid CIDs anyway, because we'd never have inserted them, so naively we could do without it. Is the intention to guard against timing attacks when a cryptographic mechanism is in use? If so, that should be noted in the docs so future readers aren't confused about the benefit.

File: quinn-proto/src/cid_generator.rs
Comment (Ralith) [inline quinn-proto/src/cid_generator.rs:9]: We should have a brief note here describing desirable properties for connection IDs, e.g. that they be indistinguishable from random data by an external viewer.

**Review by Ralith — CHANGES_REQUESTED** [review]
Thanks for the updates! Made a deep pass; mostly cosmetic issues.

File: quinn-proto/src/shared.rs
Comment (liwenjieQu) [reply quinn-proto/src/shared.rs:65]: Should be enough to use `ConnectionId::new`. I am reverting the above changes.

Comment (liwenjieQu) [reply quinn-proto/src/shared.rs:65]: Yes. I think it is enough to use `ConnectionId::new`. I am reverting the above changes.

File: quinn-proto/src/cid_generator.rs
Comment (liwenjieQu) [reply quinn-proto/src/cid_generator.rs:9]: I will add some notes from RFC doc on Connection ID properties

File: quinn-proto/src/cid_generator.rs
Comment (liwenjieQu) [reply quinn-proto/src/cid_generator.rs:11]: > It looks like this is only used to skip checking an Endpoint's hash table for a CID, but that hash table should return None for invalid CIDs anyway, because we'd never have inserted them, so naively we could do without it.
Yes I have noticed that hash table will guarantee that non-existing CID will remain invalid. The purpose of validating CID is indeed based on some security concern. 

The major purpose of this API is to limit the amount of time for which a CID is valid in order to reduce the number of valid IDs that could be accumulated by an attacker. I am pushing another PR later (where peer actively sends NEW_CONNECTION_ID frame to rotate the CID). As a result, `fn validate_cid` is used to mark any CID that may still appear in hash table but considered invalid by application layer logic. 

This seems like an optional API. Thus, I am providing a default implementation and more docs.

Comment (liwenjieQu) [reply quinn-proto/src/cid_generator.rs]: And I think timing attack is a good example on what may happen if an attacker holds a number of valid CIDs for a long time.

File: quinn-proto/src/cid_generator.rs
Comment (Ralith) [inline quinn-proto/src/cid_generator.rs:13]: Formatting error:
```suggestion
    /// an external observer (that is, one that does not cooperate with the
    /// issuer) to correlate them with other connection IDs for the same
    /// connection.
```

File: quinn-proto/src/cid_generator.rs
Comment (Ralith) [inline quinn-proto/src/cid_generator.rs:31]:
```suggestion
/// Generates purely random connection IDs of a certain length
```

File: quinn-proto/src/lib.rs
Comment (Ralith) [inline quinn-proto/src/lib.rs:63]: We should reexport `RandomConnectionIdGenerator` to allow easy customization of connection ID length.

File: quinn-proto/src/cid_generator.rs
Comment (Ralith) [inline quinn-proto/src/cid_generator.rs:50]: We can avoid a heap allocation here:
```suggestion
        let mut bytes_arr = [0; MAX_CID_SIZE];
```

File: quinn-proto/src/cid_generator.rs
Comment (Ralith) [inline quinn-proto/src/cid_generator.rs:21]: Stray whitespace. Unless I'm missing something about markdown?
```suggestion
    /// but considered invalid by application-layer logic.
    /// e.g. we may want to limit the amount of time for which a CID is valid
    /// in order to reduce the number of valid IDs that could be accumulated
    /// by an attacker.
```

File: quinn-proto/src/cid_generator.rs
Comment (Ralith) [inline quinn-proto/src/cid_generator.rs:24]: Good call on the default implementation! I'd still like to understand the problem this is solving in more detail, though. Quinn already limits the number of connection IDs that may be outstanding per connection via the mechanisms discussed in [Issuing Connection IDs](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-issuing-connection-ids), enforced [inside `CidQueue`](https://github.com/djc/quinn/blob/master/quinn-proto/src/cid_queue.rs#L41-L43), currently hardcoded to at most 4. This is communicated to the peer via the `active_connection_id_limit` transport parameter, which helps ensure that well-behaved peers don't encounter issues.

Is there a CID exhaustion attack that this does not cover? If so, should we pursue a mechanism that also cleans them out of the hash table proactively, so as to avoid memory exhaustion?

**Review by Ralith — CHANGES_REQUESTED** [review]

File: quinn-proto/src/cid_generator.rs
Comment (liwenjieQu) [reply quinn-proto/src/cid_generator.rs:48]: Sounds good

File: quinn-proto/src/cid_generator.rs
Comment (liwenjieQu) [reply quinn-proto/src/cid_generator.rs:24]: > Is there a CID exhaustion attack that this does not cover? 

i agree that for well-behaved peers we shouldn't expect issues. quinn has followed RFC to protect each pair of communications.

This `fn validate_cid`  is a precaution to avoid when an attack can somehow eardrop `multiple` pairs of communications and accumulate many CIDs if server doesn't proactively invalidate them. As a result, I will suggest peer (especially server side) can push `NEW_CONNECTION_ID` with updated `retire_prior_to` field  (currently, quinn always set it to 0) to rotate CID. This will somehow reduce the risk of "CID exhaustion attack". I have already finished the code on my local branch.

A well-behaved peer should reply with `RETIRE_CONNECTION_ID` to complete the CID rotation. If not,  `validate_cid` then will kick in to stop the server to keep using the retired ones. 

>If so, should we pursue a mechanism that also cleans them out of the hash table proactively, so as to avoid memory exhaustion?

I think it will be good if we can clean those CIDs that are invalidated. Could be a future improvement. 👍

Comment (Ralith) [reply quinn-proto/src/cid_generator.rs:24]: Oh, I see, the real goal here is CID key rotation or similar. That's definitely something we want to support. My concern is that the approach currently taken here will place well-behaved peers at risk of connection loss due to not employing `retire_prior_to`. Additionally, there's no mechanism introduced to issue new CIDs, so a long-lived connection will run out and be lost. Therefore I don't think `validate_cid` is useful as written.

Since it sounds like you have a better approach almost ready to go, can we `validate_cid` here, merge the rest, and address CID rotation in a future PR?

Comment (Matthias247) [reply quinn-proto/src/cid_generator.rs:24]: > Since it sounds like you have a better approach almost ready to go, can we omit validate_cid here, merge the rest, and address CID rotation in a future PR?

Wouldn't this break users if the start implementing the new trait, and the additional method has to be added later on?
If it's forseeable that it will be required or even that it could be useful there shouldn't be a lot of harm with adding it - even if it's not called right now. The implementation should be straightforward for most CID formats.

Comment (liwenjieQu) [reply quinn-proto/src/cid_generator.rs:24]: > My concern is that the approach currently taken here will place well-behaved peers at risk of connection loss due to not employing retire_prior_to

Totally understood. I double check the description in [RFC](https://www.ietf.org/id/draft-ietf-quic-transport-31.html#name-consuming-and-retiring-conn) where I quote 
> "Upon receipt of an increased Retire Prior To field, the peer MUST stop using the corresponding connection IDs and retire them with RETIRE_CONNECTION_ID frames before adding the newly provided connection ID to the set of active connection IDs."

Since peer `MUST` has a reply, I don't think if a peer who doesn't employ the `retire_prior_to` can be categorized as "well-behaved" (at least it is not RFC-compliant). 

The `validate_cid` function provides a API/approach to apply what is mentioned in RFC where
> "The endpoint SHOULD continue to accept the previously issued connection IDs until they are retired by the peer. If the endpoint can no longer process the indicated connection IDs, it MAY close the connection."

Since it is a `MAY` in RFC, the CID generator creator can thus decide whether/when we want to close the connection based on their own implementation. In the current PR, the false result of `validate_cid` will only ignore the packets from peer, which is not exact equivalent to closing the connection. This could be a future improvement.

With that said, 
> Since it sounds like you have a better approach almost ready to go

I have a PR ready that implements the advance of `retire_prior_to` but the not form of cid generator trait.

Since the current calling of `validate_cid` is not perfect yet. I am willing to rollback this API. Regarding to Matthias's concern, it is also feasible to add it back in the future without breaking compatibility if we provide a default implementation as I have now: users can update their implementation if they want to use this feature.

Comment (Ralith) [reply quinn-proto/src/cid_generator.rs]: > Since peer MUST has a reply, I don't think if a peer who doesn't employ the retire_prior_to can be categorized as "well-behaved" (at least it is not RFC-compliant).

Right, the problem is that quinn itself does not yet generate `retire_prior_to`, hence if any implementer ever returned `false` from `validate_cid` under current conditions it would necessarily cause a CID that might still be legally used by the peer to be ignored.

> If it's forseeable that it will be required or even that it could be useful there shouldn't be a lot of harm with adding it

Ideally we'd want to tightly couple the local invalidation of CIDs with the issuing of a `retire_prior_to` and (deferred) updates to the Endpoint's hash table so that everything is guaranteed to stay in sync, which will likely require a different interface.

**Review by Ralith — APPROVED** [review]
Thanks for your work!

**Comment by liwenjieQu** [comment]
Appreciate your time @Ralith !

**Comment by liwenjieQu** [comment]
Could you squash and merge these commits so I can push a new PR? Thanks

**Comment by Ralith** [comment]
I'd prefer to get a signoff from @djc first. You can make a new PR based on this branch, and when we merge this to mater you should be able to rebase the new PR onto master without issue, and I'll happily review the new parts in the mean time. You could also squash this branch into a single commit yourself first for extra insurance against git getting tangled up.

**Comment by liwenjieQu** [comment]
Just realize that I need to remove 'crate' from `pub(crate)` before new function for ConnectionId, and thus a third-party cid generator trait can implement `generate_cid` on its own

**Comment by Ralith** [comment]
Whoops, good catch.

**Review by Ralith — APPROVED** [review]

**Comment by liwenjieQu** [comment]
The error from CI is mysterious. I am using a Mac and all tests pass on my side.

**Comment by Ralith** [comment]
There's a similar issue on #850; I'm guessing it's on github's end and we should ignore it.

**Comment by liwenjieQu** [comment]
Please not merge the PR now. I need to update the CID generator trait for proactive cid rotation https://github.com/liwenjieQu/quinn/pull/1 where the lifetime/duration of a CID needs to be returned. I think it is better to declare in the very beginning to avoid breaking compatibility

**Comment by liwenjieQu** [comment]
Sorry for the inconvenience that the PR has to be reviewed again.

File: quinn-proto/src/cid_generator.rs
Comment (Ralith) [inline quinn-proto/src/cid_generator.rs:17]:
```suggestion
    /// In the future, connection IDs will be retired after the returned `Duration`, if any.
```

**Review by Ralith — CHANGES_REQUESTED** [review]
Good catch. Minor API breakage isn't the end of the world at present, but might as well avoid it when we can.

**Review by Ralith — APPROVED** [review]

**Comment by djc** [comment]
Sorry for the delayed review -- thanks for working on this, it looks pretty good!

File: quinn-proto/src/endpoint.rs
Comment (infinity0) [inline quinn-proto/src/endpoint.rs:348]: @liwenjieQu Is there any reason we hardcode the generator here, and don't use the `local_cid_generator`? The main difference I see between `RandomConnectionIdGenerator` vs `local_cid_generator` is the `cid_len`.

Here we pass `remote_id` to `add_connection`. However inside that function, we pass `self._local_cid_generator` to `Connection::new` which uses its length, instead of the length of `remote_id` (which was generated by `RandomConnectionIdGenerator`).

Should we change this to be `self.local_cid_generator.generate_cid()` for consistency.?

Comment (Ralith) [reply quinn-proto/src/endpoint.rs:348]: Note the name `remote_id`; we're generating a *destination* connection ID here, not a local one. The client uses a random destination connection ID for its initial flight of packets; see https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-negotiating-connection-ids for details.

Comment (infinity0) [reply quinn-proto/src/endpoint.rs:348]: @Ralith Thanks for the explanation! That clears it up.

**Pull request merged**
